### DISCUSSION TITLE
Reading CSV with multiple coulmn and rows

### DISCUSSION THREAD

**Discussion by mvkpenn22** [idaholab/moose#26369] [Q&A Modules: Solid mechanics]
Hi 

I want to read multiple columns (3 columns) and multiple rows (3 rows) for youngs_modulus, my block id is 0 and data is constant over the whole block. The code I have written for one value reading from the CSS file is working fine for this. How to do for multiple columns and multiple rows.
I’m struggling with this

```
[Functions]
    [bl]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = ‘0’
    []
  []
  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E.csv'
      read_type = 'block'
      nprop = 1
      nblock = 1
    []
  []

  [Materials]
    [a]
      type = GenericFunctionMaterial
      block = 0
      prop_names = youngs_modulus
      prop_values = bl
    []
    [compute_stress]
      type = ComputeIsotropicElasticityTensor
     youngs_modulus = E
    []
```
For this code, E.csv looks like this
```
0, 10000.0
```

but for my project, it should be like this
```
0, 10000.0, 10, 10000.0
0, 30000.0, 10, 40000.0
0, 50000.0, 10, 40000.0

```

0 represents the block id, as this data remains same for whole block. I am struggling how to incorporate this

**Comment by lindsayad** [comment]
What do columns 2 and 3 represent here? A different block ID (10) and associated Young's modulus?

Also, what do rows 1 and 2 represent? (I am using 0-based numbering in my response)

**Reply by GiudGiud** [reply]
Rows should be different subdomains in that "read_type" for the PropertyFileReader

Columns should be different properties.

**Reply by GiudGiud** [reply]
So:
> multiple columns (3 columns) and multiple rows (3 rows) for youngs_modulus

this does not make sense. Young modulus should be only one column (it can be multiple rows for different subdomains)

**Reply by mvkpenn22** [reply]
Hi @GiudGiud, @lindsayad 

I have modified my material property, and it's adjusted in a way that it's a 3x3 matrix form. I am doing a calculation (3x3) with strain(3 components) to get strain.

I have defined it in this way only.

1st column represents the block_id of the mesh and the rest (3x3) represents material property which will remain constant over the whole domain

**Reply by GiudGiud** [reply]
So the young modulus is a tensor now?

So you should have 1 row per block id, and now 10 columns:
1 for the index, 1 for each component of the matrix

then downstream of reading the file, you ll re-form the matrix component by component

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 

Yes, young's modulus is tensor (3x3), this is 3x3 Young's modulus matrix, I added 0 because of block_id

E: 
10000.0, 10, 10000.0
30000.0, 10, 40000.0
50000.0, 10, 40000.0

**Reply by mvkpenn22** [reply]
Hi @GiudGiud  

I tried this way, although it didn't show any error. But, I don't how to assign it to youngs_modulus in this matrix form. For first iteration this is 3x3, but for future this dimension can vary
```
[Samplers]
  [sample]
    type = CSVSampler
    samples_file = 'E.csv'
    column_names = 'a b c'
    execute_on = 'initial timestep_end'
  []
[]

[VectorPostprocessors]
  [data]
    type = SamplerData
    sampler = sample
    execute_on = 'initial timestep_end'
  []
[]
```

```
E.csv:
a,b,c
10000.0, 10, 10000.0
30000.0, 10, 40000.0
50000.0, 10, 40000.0
```

**Reply by GiudGiud** [reply]
You are not trying to sample. So no you should not use a CSVSampler

**Reply by manikumd22** [reply]
Hi @GiudGiud 

Then how could I do that

**Reply by GiudGiud** [reply]
Do that:
```
So you should have 1 row per block id, and now 10 columns:
1 for the index, 1 for each component of the matrix

then downstream of reading the file, you ll re-form the matrix component by component
```

use a single PropertyFileReader to read the file then 9 functions to reach each column
then in a material property, re-create the young modulus tensor with the 9 functions

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I got your point

but how do we specifically mention the column number to it? For example, I made 10 columns in E.csv for block 0 and changed nprop to 10, as now we have 10 columns. How we create 9 functions specifically mentions that the 4th function will read the 4th column from bl because in gfm I didn't find anything about the column number, we are just saying prop_values = bl, so how will take care of that?

[Functions]
    [bl]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = ‘0’
    []
  []
  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E.csv'
      read_type = 'block'
      nprop = 10
      nblock = 1
    []
  []

  [Materials]
    [a]
      type = GenericFunctionMaterial
      block = 0
      prop_names = youngs_modulus
      prop_values = bl
    []
    [compute_stress]
      type = ComputeIsotropicElasticityTensor
     youngs_modulus = youngs_modulus
    []

**Reply by GiudGiud** [reply]
```
[Functions]
[bl]
type = PiecewiseConstantFromCSV
read_prop_user_object = 'reader_block'
read_type = 'block'
column_number = ‘0’
[]
[bl_1]
type = PiecewiseConstantFromCSV
read_prop_user_object = 'reader_block'
read_type = 'block'
column_number = ‘1’
[]
...
```

**Reply by manikumd22** [reply]
Hi @GiudGiud 

It is working, Thanks

I have a question, for his iteration, I am working with a 3x3 tensor. In future iterations, I have to work on a 64x64 tensor, i guess this method of defining function 64 times is not optimized. Is there any other method to read 64 x 64 in an optimized way

**Reply by GiudGiud** [reply]
yes 64x64 will be wayy too many functions.
But you can just move the little bit of code in the PiecewiseConstantFromCSV to your material. That way you can bypass the need for a function

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 

is it possible through vectorpostprocessor technique?

**Reply by GiudGiud** [reply]
You could implement it that way too if you want

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 
Can you please share some code examples for that technique and how it can be used to assign it to material property?

for example
Initialization a nxn tensor and then creating a column tuple for each column from the CSV file using vectorpostprocessor



For example, I have tried this by creating different functions. Is there any other way to avoid defining multiple functions and in material prop_value will directly read column values using VectorPostprocessors?
```
[Functions]
  [bl_1]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '0'
  []
  [bl_2]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '1'
  []
  [bl_3]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '2'
  []
[]

[UserObjects]
  [reader_block]
    type = PropertyReadFile
    prop_file_name = 'E_values.csv'
    read_type = 'block'
    nprop = 3
    nblock = 1
  []
[]

[VectorPostprocessors]
  [vpp]
    type = MaterialVectorPostprocessor
    material = 'mat'
    block = 0
  []
[]


[Materials]
  [mat]
    type = GenericFunctionMaterial
    block = 0
    prop_names = 'lambda1 lambda2 lambda3'
    prop_values = 'bl_1 bl_2 bl_3'
  []

  [compute_stress]
    type = ComputeNeoHookeanStress
    lambda = lambda1
    mu = 6700.0
    b = 1
  []
```

**Reply by GiudGiud** [reply]
you ll need a custom object to recreate the tensor material property from the functions.
I think @ngrilli shared some code to fill the tensor in another thread

I would say start from GenericFunctionMaterial and make a GenericFunctionTensorMaterial.
Once you have it working I can take a look at it as we can add it to the framework

**Reply by mvkpenn22** [reply]
Hi @GiudGiud / @ngrilli 

I have defined it in this way, is this the correct way and will it read the values from CSV, however, to test I defined this for C_ijkl. For my project, it will not be a 6x6 tensor, it will 64x64 for custom material, 

For fourth-order tensor this is how we define it

const MaterialProperty<RankFourTensor> & _elasticity_tensor;

but I am confused how to define a 64x64 tensor


```

[Functions]
  [bl_1]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '0'
  []
  [bl_2]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '1'
  []
  [bl_3]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = reader_block
    read_type = 'block'
    column_number = '2'
  []
[]

[UserObjects]
  [reader_block]
    type = PropertyReadFile
    prop_file_name = 'E_values.csv'
    read_type = 'block'
    nprop = 3
    nblock = 1
  []
[]

[VectorPostprocessors]
  [vpp]
    type = MaterialVectorPostprocessor
    material = 'mat'
    block = 0
  []
[]


[Materials]
  [mat]
    type = GenericFunctionMaterial
    block = 0
    prop_names = 'lambda1 lambda2 lambda3'
    prop_values = 'bl_1 bl_2 bl_3'
  []

  [./elasticity_tensor]
    type = ComputeElasticityTensor
    C_ijkl = 'lambda1 lambda2 lambda3 0 0 0
              lambda1 lambda2 lambda3 0 0 0
              lambda1 lambda2 lambda3 0 0 0
              0  0 0 lambda1 0 0 0
              0  0 0 0 lambda1 0 0
              0  0 0 0 0 lambda1 0'
  [../]
```

**Reply by GiudGiud** [reply]
> but I am confused how to define a 64x64 tensor

you ll need to start developing an object to programmatically retrieve the components from the CSV file.

**Reply by mvkpenn22** [reply]
Hi @GiudGiud ,

Can't we define something like this:

const MaterialProperty<RankSixtyFourTensor> & _material_tensor;

and define the values int he input file like this

  [./elasticity_tensor]
    type = ComputeElasticityTensor
    _material_tensor = 'lambda11 lambda12 lambda13 lambda14 lambda15 lambda16 ........
             lambda21 lambda22 lambda23 lambda24 lambda25 lambda26 ........
              lambda31 lambda32 lambda33 lambda34 lambda35 lambda36 ........
              lambda41 lambda42 lambda43 lambda44 lambda45 lambda46 ........
              lambda51 lambda52 lambda53 lambda54 lambda55 lambda56 ........
              lambda61 lambda62 lambda63 lambda64 lambda65 lambda66 ........
               ................................................................................................'

**Reply by GiudGiud** [reply]
if you want to. You better write a script to write those inputs then! 64 x 64 > 3600 !

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 

I got your point,

I have started working on what you mentioned. However, I must also define the nonsquare tensor of form (nxm or nx1).
Can you please tell me how to define nxm and nx1 tensor, I didn't find anything in the documentation

This is how we define 2x2 tensor
MaterialProperty <RankTwoTensor>

**Reply by GiudGiud** [reply]
@recuero or @dschwen on non-square tensor support

**Reply by recuero** [reply]
You can use `DenseMatrix` to define tensors that are not directly supported by the tensor mechanics module, including non-square ones. See, for example, `HillElastoPlasticityStressUpdate`.

**Reply by mvkpenn22** [reply]
Hi @GiudGiud , @recuero 

I have defined DenseMatrix (_weight_tensor), that should be of form _weight_tensor(2,), but I am getting error:

Error:
In file included from /storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/build/unity_src/materials_lagrangian_Unity.C:16:
/storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/src/materials/lagrangian/ComputeNeoHookeanStress.C: In member function 'virtual void ComputeNeoHookeanStress::computeQpPK2Stress()':
/storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/src/materials/lagrangian/ComputeNeoHookeanStress.C:61:28: error: conversion from 'libMesh::DenseMatrix<double>' to non-scalar type 'RankTwoTensor' {aka 'RankTwoTensorTempl<double>'} requested
     RankTwoTensor strain = _weight_tensor;
                            ^~~~~~~~~~~~~~
                     make: *** [/storage/work/mvk6158/moose_directories/moose//framework/build.mk:151: /storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/build/unity_src/materials_lagrangian_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1


.h file:

#pragma once

#include "ComputeLagrangianStressPK2.h"

class ComputeNeoHookeanStress : public ComputeLagrangianStressPK2
{
public:
  static InputParameters validParams();
  ComputeNeoHookeanStress(const InputParameters & parameters);

protected:
  /// Actual stress/Jacobian update
  virtual void computeQpPK2Stress();

protected:
  const MaterialProperty<Real> & _lambda;
  const MaterialProperty<Real> & _mu;
  DenseMatrix<Real> _weight_tensor;
};


.C file:

#include "ComputeNeoHookeanStress.h"

registerMooseObject("TensorMechanicsApp", ComputeNeoHookeanStress);

InputParameters
ComputeNeoHookeanStress::validParams()
{
  InputParameters params = ComputeLagrangianStressPK2::validParams();

  params.addParam<MaterialPropertyName>("lambda",
                                        "lambda",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("mu",
                                        "mu",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");                                    
  return params;
}

ComputeNeoHookeanStress::ComputeNeoHookeanStress(const InputParameters & parameters)
  : ComputeLagrangianStressPK2(parameters),
    _lambda(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda"))),
    _mu(getMaterialProperty<Real>(getParam<MaterialPropertyName>("mu"))),
    _weight_tensor(2,2)
{
}

void
ComputeNeoHookeanStress::computeQpPK2Stress()
{
  if (_large_kinematics)
  {
    RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
    _S[_qp] = (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
    _C[_qp] =
        -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
        _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
  }
  // Small deformations = linear strain
  else
  {
    const auto I = RankTwoTensor::Identity();
    RankTwoTensor strain = _weight_tensor;
    _S[_qp] = strain;
  }
}

**Reply by mvkpenn22** [reply]
Hi @GiudGiud ,


I am trying to define RankTwoTensor (weightmatrix) but getting an error,


*** ERROR ***
To use fillGeneralFromInputVector, your input must have size 81. Yours has size 4[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0


why it is asking for 81 components, as I am defining RankTwoTensor which has 9 components.

So what should I change, so that I will only need 9 components? For example, if I define RankThreeTensor


.h file:
  const MaterialProperty<Real> & _lambda;
  const MaterialProperty<Real> & _mu;
  RankTwoTensor _weightmatrix;


.C File:

  params.addRequiredParam<std::vector<Real>>("weightmatrix", "weight from neural network");
  params.addParam<MooseEnum>(
      "fill_method", RankTwoTensor::fillMethodEnum() = "general", "The fill method");                                      
  return params;
}

ComputeNeoHookeanStress::ComputeNeoHookeanStress(const InputParameters & parameters)
  : ComputeLagrangianStressPK2(parameters),
    _lambda(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda"))),
    _mu(getMaterialProperty<Real>(getParam<MaterialPropertyName>("mu"))),
    _weightmatrix(this->template getParam<std::vector<Real>>("weightmatrix"),
                  (RankTwoTensor::FillMethod)(int)this->template getParam<MooseEnum>("fill_method"))
{
}

**Reply by mvkpenn22** [reply]
Hi @GiudGiud ,

I have defined DenseMatrix (_weight_matrix), and I have defined its components and then assigned them to _S[_qp] = _weight_matrix[_qp]. But I am getting an error. Can you please help me with this?

**Error:**

In file included from /storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/build/unity_src/materials_lagrangian_Unity.C:16:
/storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/src/materials/lagrangian/ComputeNeoHookeanStress.C: In member function 'virtual void ComputeNeoHookeanStress::computeQpPK2Stress()':
/storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/src/materials/lagrangian/ComputeNeoHookeanStress.C:62:15: error: no match for 'operator=' (operand types are 'MooseADWrapper<RankTwoTensorTempl<double>, false>' {aka 'RankTwoTensorTempl<double>'} and 'MaterialProperty<libMesh::DenseMatrix<double> >')
     _S[_qp] = _weight_matrix;
               ^~~~~~~~~~~~~~




.h file:

#pragma once

#include "ComputeLagrangianStressPK2.h"
#include "DenseMatrix.h"

class ComputeNeoHookeanStress : public ComputeLagrangianStressPK2
{
public:
  static InputParameters validParams();
  ComputeNeoHookeanStress(const InputParameters & parameters);

protected:
  virtual void computeQpPK2Stress();

protected:
  const MaterialProperty<Real> & _lambda;
  const MaterialProperty<Real> & _lambda1;
  const MaterialProperty<Real> & _lambda2;
  const MaterialProperty<Real> & _lambda3;
  const MaterialProperty<Real> & _lambda4;
  const MaterialProperty<Real> & _mu;
  MaterialProperty<DenseMatrix<Real>> & _weight_matrix;
};


.C file:

#include "ComputeNeoHookeanStress.h"

registerMooseObject("TensorMechanicsApp", ComputeNeoHookeanStress);

InputParameters
ComputeNeoHookeanStress::validParams()
{
  InputParameters params = ComputeLagrangianStressPK2::validParams();

  params.addParam<MaterialPropertyName>("lambda",
                                        "lambda",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("mu",
                                        "mu",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda1",
                                        "lambda1",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda2",
                                        "lambda2",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda3",
                                        "lambda3",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda4",
                                        "lambda4",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");

  return params;
}

ComputeNeoHookeanStress::ComputeNeoHookeanStress(const InputParameters & parameters)
  : ComputeLagrangianStressPK2(parameters),
    _lambda(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda"))),
    _lambda1(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda1"))),
    _lambda2(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda2"))),
    _lambda3(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda3"))),
    _lambda4(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda4"))),
    _mu(getMaterialProperty<Real>(getParam<MaterialPropertyName>("mu"))),
    _weight_matrix(declareProperty<DenseMatrix<Real>>("matrix_prop"))

{
}

void
ComputeNeoHookeanStress::computeQpPK2Stress()
{
  usingTensorIndices(i_, j_, k_, l_);
  {
    const auto I = RankTwoTensor::Identity();
    _weight_matrix[_qp].resize(2, 2);
    _weight_matrix[_qp](0, 0) = _lambda1[_qp];
    _weight_matrix[_qp](0, 1) = _lambda2[_qp];
    _weight_matrix[_qp](1, 0) = _lambda3[_qp];
    _weight_matrix[_qp](1, 1) = _lambda4[_qp];
    _S[_qp] = _weight_matrix[_qp];
  }
}

**Reply by GiudGiud** [reply]
> ress.C:62:15: error: no match for 'operator=' (operand types are 'MooseADWrapper<RankTwoTensorTempl, false>' {aka 'RankTwoTensorTempl'} and 'MaterialProperty<libMesh::DenseMatrix >')

you need to access the material property at _qp before assigning
_S[_qp] = _weight_matrix[_qp];

in general you need to make sure the types are matching when using =

**Reply by mvkpenn22** [reply]
Hi @GiudGiud ,

_S[_qp] is RankTwoTensor and _weight_matrix is the Densematrix of TwoRankTensor, I am filing the components of _weight_matrix using _lambda (constant values) and at every _qp _weight_matrix will remain the same.

Is it the wrong way to fill the values for my constant _weight_matrix?

_weight_matrix[_qp].resize(2, 2);
_weight_matrix[_qp](0, 0) = _lambda1[_qp];
_weight_matrix[_qp](0, 1) = _lambda2[_qp];
_weight_matrix[_qp](1, 0) = _lambda3[_qp];
_weight_matrix[_qp](1, 1) = _lambda4[_qp];
_weight_matrix[_qp](0, 0) = _lambda1[_qp]

**Reply by GiudGiud** [reply]
why do you have this line twice?
```
_weight_matrix[_qp](0, 0) = _lambda1[_qp]
```

otherwise this looks fine.
Here I think these
```
_lambda(getMaterialProperty(getParam("lambda"))),
```
can be like this instead:
```
_lambda(getMaterialProperty("lambda")),
```

**Reply by mvkpenn22** [reply]
Hi @GiudGiud ,


I tried with modifications, but still getting an error
```
**Error:**
In file included from /storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/build/unity_src/materials_lagrangian_Unity.C:16:
/storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/src/materials/lagrangian/ComputeNeoHookeanStress.C:76:15: error: no match for 'operator=' (operand types are 'MooseADWrapper<RankTwoTensorTempl<double>, false>' {aka 'RankTwoTensorTempl<double>'} and 'MaterialProperty<libMesh::DenseMatrix<double> >')
     _S[_qp] = _weight_matrix;
                 from /storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/build/unity_src/materials_lagrangian_Unity.C:2:
/storage/work/mvk6158/moose_directories/moose//framework/build/header_symlinks/RankTwoTensor.h:500:27: note: candidate: 'RankTwoTensorTempl< <template-parameter-1-1> >& RankTwoTensorTempl< <template-parameter-1-1> >::operator=(const RankTwoTensorTempl< <template-parameter-1-1> >&) [with T = double]'
   RankTwoTensorTempl<T> & operator=(const RankTwoTensorTempl<T> & a);
```

**Reply by GiudGiud** [reply]
did you try
```
     _S[_qp] = _weight_matrix[_qp];
```
you are reporting an error on another line as before? So maybe the other line was fixed?

**Reply by mvkpenn22** [reply]
Hi @GiudGiud ,

I tried this only _S[_qp] = _weight_matrix[_qp];



/storage/work/mvk6158/moose_directories/moose//modules/tensor_mechanics/src/materials/lagrangian/ComputeNeoHookeanStress.C:65:33: error: no match for 'operator=' (operand types are 'MooseADWrapper<RankTwoTensorTempl<double>, false>' {aka 'RankTwoTensorTempl<double>'} and 'MooseADWrapper<libMesh::DenseMatrix<double>, false>' {aka 'libMesh::DenseMatrix<double>'})
     _S[_qp] = _weight_matrix[_qp];





.h file:
protected:
  const MaterialProperty<Real> & _lambda;
  const MaterialProperty<Real> & _lambda1;
  const MaterialProperty<Real> & _lambda2;
  const MaterialProperty<Real> & _lambda3;
  const MaterialProperty<Real> & _lambda4;
  const MaterialProperty<Real> & _mu;
  MaterialProperty<DenseMatrix<Real>> & _weight_matrix;
};



.C file:
InputParameters
ComputeNeoHookeanStress::validParams()
{
  InputParameters params = ComputeLagrangianStressPK2::validParams();

  params.addParam<MaterialPropertyName>("lambda",
                                        "lambda",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("mu",
                                        "mu",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda1",
                                        "lambda1",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda2",
                                        "lambda2",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda3",
                                        "lambda3",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("lambda4",
                                        "lambda4",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  return params;
}

ComputeNeoHookeanStress::ComputeNeoHookeanStress(const InputParameters & parameters)
  : ComputeLagrangianStressPK2(parameters),
    _lambda(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda"))),
    _lambda1(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda1"))),
    _lambda2(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda2"))),
    _lambda3(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda3"))),
    _lambda4(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda4"))),
    _mu(getMaterialProperty<Real>(getParam<MaterialPropertyName>("mu"))),
    _weight_matrix(declareProperty<DenseMatrix<Real>>("matrix_prop"))

{
}

**Reply by GiudGiud** [reply]
ok so now the problem is that _weight_matrix is a denseMatrix. 
But the other one, _S, in a rank2tensor so just a 2x2 matrix.

So it's odd that they are being set equal to each other? They do not have the same size

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 
I am assigning _weight_matrix to be like this
_weight_matrix[_qp].resize(2, 2)

It is 2x2 dense matrix

For my case, I am reading and filling the dense matrix using csf file values, and I want _S to be equal to _weight_matrix

Is there any way so that _S an be assigned equal to _weight_matrix?

**Reply by GiudGiud** [reply]
ok you ll have to copy component by component.
the compiler wont understand that they are the same size

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 

Thank you so much, it's working. I am almost done with the dense matrix multiplication for my code. I want to define it in a more optimized way.

For example, I am using this to fill the DenseMAtrix with _weight_constants_layer_1

input file:

```
[Materials]

  [compute_stress]
    type = ComputeNeoHookeanStress

    weight_constants_layer_1 = "0.061217666	-0.18224742	-0.44236708	-0.23818971
                        0.39028817	0.42743036	0.1586842	0.16618702
                        -0.08433412	-0.42876166	-0.06616944	-0.28819194
                        0.2550973	-0.2711665	-0.25381505	0.16642846"
 ```

.C file:

```
 params.addRequiredRangeCheckedParam<std::vector<Real>>("weight_constants_layer_1",
                                                         "weight_constants_layer_3_size = 4096",
                                                         " for small deformations");
   
    _weight_constants_layer_1(getParam<std::vector<Real>>("weight_constants_layer_1")),
    _weight_layer_1(declareProperty<DenseMatrix<Real>>("weight_layer_1")),

    _weight_layer_1[_qp].resize(4, 64);
    for (int i = 0; i < 4; ++i) {
        for (int j = 0; j < 64; ++j) {
            _weight_layer_1[_qp](i, j) = _weight_constants_layer_1[index1];
            ++index1;
        }
    }
```


Is there any other way weight_constants_layer_1, will take 4096 values directly from the CSV file

```
weight_constants_layer_1 = 'column1 column2 ... column4096'
```

**Reply by GiudGiud** [reply]
yes, just specify a "weights_layer_1_file" instead of "weights_constant_layer_1"

then read that file and programmatically fill from the file into the tensor
